### PR TITLE
[SDK] Upgraded error messages on failed transactions

### PR DESCRIPTION
**Pull request by adam-maj**
`ethers.sendTransaction` makes an `eth_estimateGas` request before making the actual `eth_sendTransaction` request if the `gasLimit` on a transaction hasn't been manually set. If the transaction would revert, this swallows the actual error message and throws a `FAILED_TO_ESTIMATE_GAS_LIMIT` error instead.

Now, we manually call `estimateGas` ourselves to set the `gasLimit` on every transaction. If this fails, we then make an `eth_call` request to extract the actual error messages and throw a nicely formatted error.

### CODE REVIEW

**Comment by jnsdls** [comment]
huge if true

**Comment by jnsdls** [comment]
needs changeset tho @adam-maj

File: packages/sdk/src/evm/common/error.ts [hunk 1]
```diff
@@ -390,6 +384,8 @@ export async function convertToTWError(
   contractAddress: string,
   contractInterface: ethers.utils.Interface,
 ): Promise<TransactionError> {
+  console.log(error.message);
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/common/error.ts:387]: cleanup

File: packages/sdk/src/evm/common/error.ts [hunk 2]
```diff
@@ -398,9 +394,11 @@ export async function convertToTWError(
     // not sure what this is, just throw it back
     raw = error.toString();
   }
+
   let reason =
     parseMessageParts(/.*?"message[^a-zA-Z0-9]*([^"\\]*).*?/, raw) ||
     parseMessageParts(/.*?"reason[^a-zA-Z0-9]*([^"\\]*).*?/, raw);
+
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/common/error.ts:401]: weird that it strips out the `!` here, the regex looks fine

File: packages/sdk/src/evm/common/error.ts [hunk 3]
```diff
@@ -264,187 +264,114 @@ export type FunctionInfo = {
...
-    try {
-      const url = new URL(rpcUrl);
-      builtErrorMsg += withSpaces(`RPC`, url.hostname);
-    } catch (e2) {
-      // ignore if can't parse URL
+    if (info.value && info.value.gt(0)) {
+      errorMessage += withSpaces(
+        "value",
+        `${ethers.utils.formatEther(info.value)} ${
+          NATIVE_TOKENS[info.network.chainId as SUPPORTED_CHAIN_ID]?.symbol
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/common/error.ts]: || ""

File: packages/sdk/src/evm/common/error.ts [hunk 4]
```diff
@@ -264,187 +264,114 @@ export type FunctionInfo = {
...
-    this.reason = reason;
-    this.from = from;
-    this.to = to;
-    this.data = data;
-    this.chain = network;
-    this.rpcUrl = rpcUrl;
-    this.functionInfo = functionInfo;
+
+    errorMessage += `\n\n\n╔═════════════════════╗\n║ DEBUGGING RESOURCES ║\n╚═════════════════════╝\n\n`;
+    errorMessage += `Need helping debugging? Come tell us about your error in our Discord channel and we'll help you out:\nhttps://discord.gg/thirdweb`;
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/common/error.ts]: `Need helping debugging? Join our Discord https://discord.gg/thirdweb`

File: packages/sdk/src/evm/core/classes/contract-wrapper.ts [hunk 5]
```diff
@@ -395,23 +400,98 @@ export class ContractWrapper<
...
   }
 
+  private async formatError(
+    error: any,
+    fn: keyof TContract["functions"],
+    args: any[],
+    callOverrides: CallOverrides,
+  ) {
+    console.log(">>>>>>>>>>>>>>>>>>>>>>>>> RAW ERROR");
+    console.log(error);
```
Comment (joaquim-verges) [inline packages/sdk/src/evm/core/classes/contract-wrapper.ts]: remove?

**Pull request merged**